pallets/thea-staking/src/lib.rs

```diff
@@ -518,392 +669,526 @@ pub mod pallet {
-			log::trace!(target: "runtime::thea::staking", "rotating for relayers of network {:?}", network);
-			let active = Self::move_queued_to_active(network);
-			map.insert(network, active);
-			Self::compute_next_session(network, session_index);
-		}
-		// Increment SessionIndex
-		let new_session_index = session_index.saturating_add(1);
-		<CurrentIndex<T>>::put(new_session_index);
-		T::SessionChangeNotifier::on_new_session(map);
-		Self::deposit_event(Event::NewSessionStarted { index: new_session_index })
-	}
+		// Add public immutables and private mutables.
+		pub fn rotate_session() {
+			let session_index = <CurrentIndex<T>>::get();
+			log::trace!(target: "runtime::thea::staking", "rotating session {:?}", session_index);
+			let active_networks = <ActiveNetworks<T>>::get();
+			for (offender, (percent, reporters)) in <CommitedSlashing<T>>::iter() {
+				if let Some(net) =
+					active_networks.iter().find(|n| <Candidates<T>>::contains_key(n, &offender))
+				{
+					if let Some(to_slash) = <Candidates<T>>::get(net, &offender) {
+						// total amount transfered to treasury - used for reporters award
+						// calculations
+						let mut total_slashed = BalanceOf::<T>::zero();
+						let actual_percent = Percent::from_percent(percent);
+						// slashing relayer's individual stake
+						let amount: BalanceOf<T> = actual_percent * to_slash.individual;
+						// TODO: where to transfer? % > Treasury && % > to reporters
+						if <pallet_balances::Pallet<T> as Currency<_>>::transfer(
+							&offender,
+							&T::TreasuryPalletId::get().into_account_truncating(),
+							amount,
+							ExistenceRequirement::KeepAlive,
+						)
+						.is_ok()
+						{
+							total_slashed = total_slashed.saturating_add(amount);
+							Self::deposit_event(Event::Slashed { offender, amount });
+						}
+						// slash stakers / nominators
+						let nominators_count = to_slash.stakers.len();
+						if nominators_count != 0 {
+							let nominator_stake_avg = Percent::from_rational(1, nominators_count);
+							let nominators_portion =
+								to_slash.total.saturating_sub(to_slash.individual);
+							let nominator_amount_total: BalanceOf<T> =
+								nominator_stake_avg * nominators_portion;
+							let nominator_amount_individual: BalanceOf<T> =
+								actual_percent * nominator_amount_total;
+							for nominator in to_slash.stakers.iter() {
+								if <pallet_balances::Pallet<T> as Currency<_>>::transfer(
+									nominator,
+									&T::TreasuryPalletId::get().into_account_truncating(),
+									nominator_amount_individual,
+									ExistenceRequirement::KeepAlive,
+								)
+								.is_ok()
+								{
+									total_slashed =
+										total_slashed.saturating_add(nominator_amount_individual);
+									Self::deposit_event(Event::Slashed {
+										offender: nominator.to_owned(),
+										amount: nominator_amount_individual,
+									});
+								}
+							}
+						}
+						// distribute to reporters
+						let reporter_percent =
```

Comment:
So here we are predefining the co-efficient for reporter_percent, and we 
1. Calculate `reporter_award` 
```rust
let reporter_award: BalanceOf<T> = reporter_percent * total_slashed;
```
2. We just transfer this reporter award for any number of reporters 
Now, This might result in us accidentally minting or transferring more than the slashed amount, So shouldn't we do something like this 

```rust 
let reporter_total_award_part = Percent::from_percent(T::ReportersRewardCoeficient::get()); 
let reporter_total_award = reporter_percent * total_slashed; 
// All the reporters get equal weightage for rewards 
let reporter_individual_part = Percent::from_rational(1, reporters.len()); 
let reporter_individual_award = reporter_individual_part * reporter_total_award; 
for reporter in reporters.into_iter(){ 
  if <pallet_balances::Pallet<T> as Currency<<_>>::transfer( 
      &T::TreasuryPalletId::get().into_account_truncating(), 
      &reporter, 
      reporter_individual_award, 
      ExistenceRequirement::KeepAlice 
   ) 
```

Reply:
this can work if reprotrers.len() <= 100, otherwise we have to use `Permill` instead of `Percent`

Reply:
Yes, We should use `Permill`

Reply:
done